### PR TITLE
CDAP-1928,1611 Implementation of passing mapreduce counters to the Condi...

### PR DESCRIPTION
**Pull request by sagarkapare**
...tion node and execution of the condition branches based on them.

Associated Jira:
1. https://issues.cask.co/browse/CDAP-1928
2. https://issues.cask.co/browse/CDAP-1611

Hadoop counters are passed as `Map<String, String>` to the `Predicate` in the Condition node. Key of the map represents the concatenation of the counter group name and counter name by `.`, while value represents the value of the counter. 

This is currently work in progress - no unit tests yet and need more testing on the distributed cluster. 

Opening PR to get the early feedback on the approach.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 1]
```diff
@@ -265,12 +273,39 @@ private void executeNode(final ApplicationSpecification appSpec, WorkflowNode no
...
+  @SuppressWarnings("unchecked")
+  private void executeCondition(final ApplicationSpecification appSpec, WorkflowConditionNode node,
+                                final InstantiatorFactory instantiator, final ClassLoader classLoader,
+                                String previousNodeId) throws Exception {
+    Map<String, String> token = mapReduceJobCounters.get(previousNodeId);
+    if (previousNodeId == null || token == null) {
+      throw new IllegalStateException("Condition node is not followed by the MapReduce program.");
+    }
+
+    Class<?> clz = Class.forName(node.getPredicateClassName(), true, classLoader);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Seems simpler if just use `classLoader.loadClass`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 2]
```diff
@@ -265,12 +273,39 @@ private void executeNode(final ApplicationSpecification appSpec, WorkflowNode no
...
+        executeCondition(appSpec, (WorkflowConditionNode) node, instantiator, classLoader, previousNodeId);
+        break;
       default:
         break;
     }
   }
 
-  @Override
+  @SuppressWarnings("unchecked")
+  private void executeCondition(final ApplicationSpecification appSpec, WorkflowConditionNode node,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Remove final if not needed.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java [hunk 3]
```diff
@@ -32,16 +30,16 @@
...
+   * the {@link Callable#call() call} method is invoked. When the program completes,
+   * the {@link Callable} returns the {@link Map} of execution context of the program run or {@code null} if
    * no such context exists.
    * <p/>
    * <p> An Exception is thrown from the {@link Callable#call()} method if the program fails </p>
    *
    * @throws IllegalArgumentException if no program with the specified name is defined in the workflow
    */
-  Callable<RuntimeContext> getProgramRunner(String name);
+  Callable<Map<String, String>> getProgramRunner(String name);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: I think it's better to have an object to represent the execution context, even though inside that object is just a Map right now. It makes extension easier later.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java [hunk 4]
```diff
@@ -61,20 +67,32 @@
...
                                         "MapReduceProgramController");
     }
   }
+
+  private Map<String, String> getCounterMap(MapReduceContext context) throws Exception {
+    Map<String, String> mapCounters = Maps.newHashMap();
+    Counters counters = ((Job) context.getHadoopJob()).getCounters();
+    for (CounterGroup group : counters) {
+      for (Counter counter : group) {
+        mapCounters.put(group.getName() + "." + counter.getName(), Long.toString(counter.getValue()));
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java]: This assumes that neither group nor the counter name have `.` in it, right?

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java]: We are just concatenating the counter group name and counter name to represent the key in the map. We are not recreating group name and counter name from the key, so it does not matter even if the counter name has `.` in it.
e.g. assume the group name is `some.group.name` and counter name is `some.counter.name`. The key in the map will be `some.group.name.some.counter.name`.  We will document the same.

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java]: But if group name and counter name can have `.` in it, then there can be conflicts right?
For instance:
group name: `some.group.name` and counter name: `foo`.
group name: `some.group` and counter name: `name.foo`

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java]: ahh.. makes sense. will remove the `.` while concatenating.

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java]: I mean - I don't think that resolves the problem.
With those two metrics, it'll still conflict.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: using `classLoader.loadClass` now.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Removed.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: Using `WorkflowToken` to wrap the hadoop counters map.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/MapReduceProgramWorkflowRunner.java]: Updated the map definition.

**Comment by sagarkapare** [comment]
@chtyim @anwar6953 Please take a look again. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/SparkProgramWorkflowRunner.java [hunk 5]
```diff
@@ -63,20 +65,21 @@
...
     if (controller instanceof SparkProgramController) {
-      final RuntimeContext context = ((SparkProgramController) controller).getContext();
-      return executeProgram(controller, context);
+      executeProgram(controller, ((SparkProgramController) controller).getContext());
     } else {
       throw new IllegalStateException("Failed to run program. The controller is not an instance of " +
                                         "SparkProgramController");
     }
+    // TODO: currently Spark program returns the null token
+    return null;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/SparkProgramWorkflowRunner.java]: seems that this should return an empty token? Actually, the token should be created at start of the workflow. Then it is passed to every action, and the action can modify the token. No action should create a new token, or return null.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 6]
```diff
@@ -265,12 +273,39 @@ private void executeNode(final ApplicationSpecification appSpec, WorkflowNode no
...
   }
 
-  @Override
+  @SuppressWarnings("unchecked")
+  private void executeCondition(ApplicationSpecification appSpec, WorkflowConditionNode node,
+                                InstantiatorFactory instantiator, ClassLoader classLoader,
+                                String previousNodeId) throws Exception {
+    WorkflowToken token = mapReduceJobCounters.get(previousNodeId);
+    if (previousNodeId == null || token == null) {
+      throw new IllegalStateException("Condition node is not followed by the MapReduce program.");
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: you mean "not preceded by a MapReduce"?

File: cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  static final class ConditionalWorkflow extends AbstractWorkflow {
+
+    @Override
+    protected void configure() {
+      setName("ConditionalWorkflow");
+      setDescription("Workflow to test Condition nodes");
+      addMapReduce("RecordVerifier");
+      condition(new MyVerificationPredicate()).addMapReduce("ClassicWordCount").end();
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java]: no like this syntax. seems ambiguous to me. 
a) should be addCondition(). or ifCondition(). 
b) should be followed thenDo()...elseDo()...done();
Not sure what the right syntax is, but this one seems counter-intuitive.

Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java]: I think in this example, there is no else part. The complete syntax looks something like:

```java
condition(Predicate)
  .addMapReduce()
  .addMapReduce()
.otherwise()
  .addMapReduce()
.done()
```

there is no `then` because it is unnecessary and also more align with what the `fork-join` does as well as java syntax `if () { } else { }`.

The naming of the keywords `condition`, `otherwise` and `done` can be changed to whatever is more clear.

Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java]: ok

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java (hunk 6 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Yeah right. Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/SparkProgramWorkflowRunner.java (hunk 5 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/SparkProgramWorkflowRunner.java]: Returning `null` here, since currently the `WorkflowToken` interface only has single method `getMapReduceCounters` in it which is specific to the MapReduce programs.

**Comment by sagarkapare** [comment]
This PR is ready for the review. Build is running here - http://builds.cask.co/browse/CDAP-DUT1423-6. @anew Please take a look. Thanks!

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java [hunk 8]
```diff
@@ -27,24 +27,33 @@
...
+   * the {@link Callable} returns the {@link WorkflowToken} for the program run or {@code null} if
    * no such context exists.
    * <p/>
    * <p> An Exception is thrown from the {@link Callable#call()} method if the program fails </p>
    *
    * @throws IllegalArgumentException if no program with the specified name is defined in the workflow
+   * @throws UnsupportedOperationException if it is called from {@link Predicate}
    */
-  Callable<RuntimeContext> getProgramRunner(String name);
+  Callable<WorkflowToken> getProgramRunner(String name);
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: I don't understand. Before, the program runner had access to the runtime arguments. Now it does not? How will runtime arguments be passed to programs then?

**Comment by anew** [comment]
@sagarkapare please implement what we discussed in person:
- the token is created at the start of the workflow
- it is passed to every action, and actions can modify it
- for now, only map/reduce can modify it, others can only read it
- in the future we can add methods to allow adding properties by any action
This means the  program runner is not a Callable but a Runnable. 

I will review this asap as soon as you are ready. Thanks

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java (hunk 8 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: Creating `WorkflowToken` in the `WorkflowDriver` and passing it along to the programs running under it.

**Comment by sagarkapare** [comment]
@anew Addressed the review comments. Please take a look again. Thanks!

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT1423-7

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java [hunk 9]
```diff
@@ -27,24 +26,29 @@
 
   WorkflowSpecification getWorkflowSpecification();
 
+  /**
+   * @throws UnsupportedOperationException if it is called from {@link Predicate}
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: why?

File: cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java [hunk 10]
```diff
@@ -40,9 +41,10 @@
 
   private final String name;
   private String programName;
-  private Callable<RuntimeContext> programRunner;
+  private Runnable programRunner;
   private SchedulableProgramType programType;
   private WorkflowContext context;
+  private WorkflowToken token;
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: why is this needed here? It also seems that you have a getter for the token, but it never gets set.

Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: Don't you get a warning in your IDE about this? Mine warns me: token is never assigned. It also warns me that the context is assigned but never used.

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java (hunk 9 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: Since this is the part of the `WorkflowContext`, it is possible that the user might invoke this method through the predicate. Per discussion with @chtyim, throwing `UnsupportedOperationException` in that case.

**Comment by anew** [comment]
Some minor comments, and I also have a more general question: 

So, the token is a class variable. In a fork, both/all branches are executed concurrently, so they all share the same token. Is that good? Or would it be better to make a copy of the token for each branch, and merge the tokens when all branches have completed. (Merge here can mean to take an arbitrary one, but when we add more capabilities to manipulate the token, it will have to become more sophisticated). 

If two branches share the same token, and both have a MR action followed by a condition, the expectation is that each condition sees the counters of the MR in its own branch. That is not guaranteed if they share a single token.

File: cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java (hunk 10 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: oops... sorry!! this was missed by me. We do not need `WorkflowToken` here and context too. Removed both.

**Comment by sagarkapare** [comment]
Yeah. Makes sense. Making deep copy of the `WorkflowToken` and passing the copied object to the each branch of the fork. Once the all branches of the fork are executed, currently as a part of the merge process taking the MapReduce counters from the first branch and assigning it to the merged token.

**Comment by sagarkapare** [comment]
@anew Please take a look again. Thanks!

File: cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java [hunk 11]
```diff
@@ -81,14 +78,14 @@ public void initialize(WorkflowContext context) throws Exception {
...
+      programRunner.run();
 
       // TODO (terence) : Put something back to context.
 
       LOG.info("{} Program {} workflow action completed",
                programType != null ? programType.name() : null, programName);
     } catch (Exception e) {
-      LOG.info("Failed to execute {} Program {} in workflow: {}", 
+      LOG.info("Failed to execute {} Program {} in workflow: {}",
                programType != null ? programType.name() : null, programName, e);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: I think `LOG` supports null value, you don't need the `?:` (if you don't call `.name()`, which is the same as `enum.toString()`.

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: Meaning just write it as `LOG.info("....", programType, programName, e);`

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: and by the way, when logging with an exception, you should log the stacktrace as well. Meaning it should be:

`LOG.info("Failed to execute {} Program {} in workflow", programType, programName, e);    // Remove the last {}`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Implementation of the {@link WorkflowToken} interface.
+ */
+public class BasicWorkflowToken implements WorkflowToken {
+  private Map<String, Map<String, Long>> mapReduceCounters;
+
+  @Nullable
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Do we need to allow null? Can it be an empty map if there is no counter? Or we do need to use `null` to distinguish no counter vs empty counter?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+  public void setMapReduceCounters(Map<String, Map<String, Long>> mapReduceCounters) {
+    this.mapReduceCounters = mapReduceCounters;
+  }
+
+  /**
+   * Make a deep copy of the WorkflowToken. Currently only copies the MapReduce counters.
+   * @param token WorkflowToken to be copy
+   * @return copied WorkflowToken
+   */
+  public static WorkflowToken deepCopy(WorkflowToken token) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Why a static method? Why not an instance method that creates a clone of itself?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    BasicWorkflowToken copiedToken = new BasicWorkflowToken();
+    if (token.getMapReduceCounters() != null) {
+      Map<String, Map<String, Long>> copiedMap = new HashMap<String, Map<String, Long>>(token.getMapReduceCounters());
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Is this a deep copy?

Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Fixed. My bad.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 13 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Made instance method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 12 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: We don't use `null` to distinguish. Simpler if we store empty map for no counter. Updated the code to fix it accordingly.

File: cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java (hunk 11 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/internal/workflow/ProgramWorkflowAction.java]: Fixed.

**Comment by sagarkapare** [comment]
@chtyim @anew Addressed review comments. Please take a look. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 12 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: I would prefer it to return null if it is not set. The reason is that in the future we want to allow setting and retrieving arbitrary properties, not just M/R counters. If a property does not exist, it will return null. It should be the same for the M/R counters (they will be serialized into a property, too).

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java (hunk 9 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java]: ok

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Interface to represent the data that is transferred from one node to the next node in the {@link Workflow}
+ */
+public interface WorkflowToken {
+  /**
+   *
+   * @return the Hadoop MapReduce counters from the previous MapReduce program in the Workflow
+   */
+  public Map<String, Map<String, Long>> getMapReduceCounters();
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java]: I think this should be @Nullable if the counters are not set in the token. And the javadoc should mention what it returns (null or empty map, I vote for null).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.common.collect.Maps;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Implementation of the {@link WorkflowToken} interface.
+ */
+public class BasicWorkflowToken implements WorkflowToken {
+  private Map<String, Map<String, Long>> mapReduceCounters = Maps.newHashMap();
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: see my other comment, it should not an empty map.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Override
+  public Map<String, Map<String, Long>> getMapReduceCounters() {
+    return mapReduceCounters;
+  }
+
+  public void setMapReduceCounters(Map<String, Map<String, Long>> mapReduceCounters) {
+    this.mapReduceCounters = mapReduceCounters;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: you should make a defensive copy into an Immutable map here, because you can't know whether the caller will modify the map later-on.

**Comment by anew** [comment]
I think almost everything is good now, but please see my comments on @Nullable getMapReduceCounters() and defensive copy on setMapReduceCounters().

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 12 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Agree. We need to document that clearly to avoid causing unintended NPE :)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  /**
+   * Make a deep copy of the WorkflowToken. Currently only copies the MapReduce counters.
+   * @param token WorkflowToken to be copy
+   * @return copied WorkflowToken
+   */
+  public WorkflowToken deepCopy(WorkflowToken token) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Shouldn't take any parameter. It should create a copy of "this" (itself).

File: cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java (hunk 15 above)
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowToken.java]: Made it @Nullable and added javadoc

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 16 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 17 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Updated the code per comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java (hunk 18 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java]: Fixed.

**Comment by sagarkapare** [comment]
@anew Please take a look again. Thank you!

**Comment by anew** [comment]
I have two more comments...

**Comment by sagarkapare** [comment]
@anew Fixed the code per suggestions. Please take a look again. Thank you!

**Comment by anew** [comment]
LGTM

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT1423-11

**Comment by sagarkapare** [comment]
Build passed here - http://builds.cask.co/browse/CDAP-DUT1423-JOB1-12 Merging.
